src/main/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfig.java

```diff
@@ -0,0 +1,405 @@
+                                @Nullable String tableNameField, @Nullable String tableName, String clientId,
+                                String clientSecret, String restApiEndpoint, String user, String password,
+                                @Nullable String valueType, @Nullable String startDate, @Nullable String endDate) {
+    this.referenceName = referenceName;
+    this.queryMode = queryMode;
+    this.applicationName = applicationName;
+    this.tableNameField = tableNameField;
+    this.tableName = tableName;
+    this.clientId = clientId;
+    this.clientSecret = clientSecret;
+    this.restApiEndpoint = restApiEndpoint;
+    this.user = user;
+    this.password = password;
+    this.valueType = valueType;
+    this.startDate = startDate;
+    this.endDate = endDate;
+  }
+
+  public String getReferenceName() {
+    return referenceName;
+  }
+
+  public SourceQueryMode getQueryMode(FailureCollector collector) {
+    SourceQueryMode mode = getQueryMode();
+    if (mode != null) {
+      return mode;
+    }
+
+    collector.addFailure("Unsupported query mode value: " + queryMode,
+      String.format("Supported modes are: %s", SourceQueryMode.getSupportedModes()))
+      .withConfigProperty(PROPERTY_QUERY_MODE);
+    throw collector.getOrThrowException();
+  }
+
+  public SourceQueryMode getQueryMode() {
+    Optional<SourceQueryMode> sourceQueryMode = SourceQueryMode.fromValue(queryMode);
+
+    return sourceQueryMode.isPresent() ? sourceQueryMode.get() : null;
+  }
+
+  public SourceApplication getApplicationName(FailureCollector collector) {
+    SourceApplication application = getApplicationName();
+    if (application != null) {
+      return application;
+    }
+
+    collector.addFailure("Unsupported application name value: " + applicationName,
+      String.format("Supported applications are: %s", SourceApplication.getSupportedApplications()))
+      .withConfigProperty(PROPERTY_APPLICATION_NAME);
+    throw collector.getOrThrowException();
+  }
+
+  @Nullable
+  public SourceApplication getApplicationName() {
+    Optional<SourceApplication> sourceApplication = SourceApplication.fromValue(applicationName);
+
+    return sourceApplication.isPresent() ? sourceApplication.get() : null;
+  }
+
+  @Nullable
+  public String getTableNameField() {
+    return tableNameField;
+  }
+
+  @Nullable
+  public String getTableName() {
+    return tableName;
+  }
+
+  public String getClientId() {
+    return clientId;
+  }
+
+  public String getClientSecret() {
+    return clientSecret;
+  }
+
+  public String getRestApiEndpoint() {
+    return restApiEndpoint;
+  }
+
+  public String getUser() {
+    return user;
+  }
+
+  public String getPassword() {
+    return password;
+  }
+
+  public SourceValueType getValueType(FailureCollector collector) {
+    SourceValueType type = getValueType();
+    if (type != null) {
+      return type;
+    }
+
```

Comment:
We need to handle the case where value type is a macro.

Reply:
Done